### DISCUSSION TITLE
Running tasks in Docker containers with ease

### DISCUSSION THREAD

**Discussion by elongl** [apache/airflow#22384] [Q&A]
I have the constraint that my tasks are isolated so that there couldn't be conflicts between tasks in such a way that it'd be impossible that multiple tasks would attempt to bind the same port, or write to the same path on the file system, etc.

My current solution is to use the [TaskFlow With Docker](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html#using-the-taskflow-api-with-docker-or-virtual-environments).

I instructed all my users to use my own custom decorator:
```py
def docker_task(task_func, *args, **kwargs):
    return task.docker(
        task_func,
        image='apache/airflow:2.2.4',
        auto_remove=True,
        network_mode='airflow_default',
        environment={k: v for k, v in os.environ.items()
                     if k.startswith('AIRFLOW')},
        container_name=f'airflow-task-{task_func.__name__}',
        *args,
        **kwargs,
    )

@docker_task
def sample_task():
    print('Hello!')
```
I'm running the container within the Airflow network (`network_mode='airflow_default'`),
and placing Airflow's environment variables so that it knows where the various services are, such as `AIRFLOW__CORE__SQL_ALCHEMY_CONN`.

Although it works, it feels to me a bit _hacky_ and I was wondering if this is relatively future-proof or whether anyone has a better way of achieving my goal.

Also, I'm still facing the issue that I cannot import my own Python modules as they are not passed to the container.

```py
import my_own_module

@docker_task
def sample_task():
    my_own_module.func() # <-- This would fail (ImportError).
```

I can write my own package and build a custom image that installs it so non of the imports are local, but that's quite a burden that I'd rather avoid.

If someone tackled in general such a constraint and could share how they solved it, it'll be appreciated.

Thanks.

**Comment by potiuk** [comment]
@ashb -> I am not sure if this was initially foreseen in @task.docker, but that sounds like reasonable behaviour to bring the environment of airflow to the docker-decorated task (to make it equivalent to any other task for interacting with Airflow) - but maybe there is another way of doing it, which I am not aware?  and mabe we should make it a default? 

Also --maybe we should also make the airflow config part of this (i.e. to find a way to simply make sure get_conf() in the  docker-decorated task is the same as in Airflow? Not sure what was the initial thinking behind it so worth discussing it at least. 

I think the "module" sharing is not "easy" to do - especially that DAGs might be present in a volume that might simply be not present in the docker engine (depending if it is a local or remote engine for example). So we cannot really mount it there in all cases, But @elongl - in your specific case if you do use a shared volume to keep dags (or when you use LocalExecutor) and if you can mount the dag folder, then I think it will work by adding appropriate mounts to be in the same place in the container as you have them as long as PYTHONPATH will be set properly. But unlike configuration distribution, making it "common" acrross all the possible executors and configurations of shared dags might be impossible or very tricky.

**Reply by elongl** [reply]
So regarding sharing the modules, you suggest that building an image that contains all the dependencies would be the simpler choice, am I right?

**Reply by potiuk** [reply]
Yeah. And more portable for all configurations (but maybe for your particular case, mounting DAGs to inside the container will do the trick) - LocalExecutor + Local Docker Engine is likely to be easy to do by adding a DAG_FOLDER -> DAG_FOLDER mount and making sure PYTHON_PATH includes it (if those modules are available as shared code in DAGs that is).

**Reply by elongl** [reply]
Yeah, I ended up building an image that includes my code and adds it to the `PYTHONPATH`.  
Seems to work fine.

I decided to lay off the `task.docker` altogether as it causes much more trouble and confusion, and simply use `DockerOperator`.

I seriously think this feature should either be removed or improved, but at it's current state it's not really solid.

**Reply by potiuk** [reply]
Improvement is an option. I tihnk it is pretty solid for cases where you just need to run Python code in isolation without using any Airlfow packages and importing extra code (again you seem to forget that not always "your need" is the same as "others need"). So it has it's good use. You are told from very beginning it has limitations - and contributions to improve it are always welcome.
